File: 61-rotate-list/rotate-list.java
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        ListNode temp = new ListNode();
        temp = head;
        if(head==null || head.next==null || k==0) return head;
        int ctr = 1;
        while(temp.next!=null){
            temp = temp.next;
            ctr++;
        }
        temp.next = head;
        k = k%ctr;
        k = ctr-k;
        while((k--) >0){
            temp = temp.next;
        }
        head = temp.next;
        temp.next = null;
        return head;
    }
}
// public static ListNode rotateRight(ListNode head, int k) {
// 	if(head==null)
//         return null;
// 	int size = 1; // since we are already at head node
// 	ListNode fast=head;
// 	ListNode slow = head;
	
// 	while(fast.next!=null){
// 	    size++;
// 	    fast = fast.next;
// 	}
	
// 	for(int i=size-k%size;i>1;i--) // i>1 because we need to put slow.next at the start.
// 		slow = slow.next;
	
//     // No dummy variable.
// 	fast.next = head;
// 	head = slow.next;
// 	slow.next = null;
	
// 	return head;
// }

// // class Solution {
// //     public ListNode rotateRight(ListNode head, int k) {
// //         if( k == 0 || head ==null || head.next == null) return head;

// //         int len=1;
// //         ListNode cur=head;
// //         //ctr and travel to the last node;
// //         while(cur.next != null){
// //             len++;
// //             cur=cur.next;
// //         }
// //         //when k=len we get the same linked list
// //         k = len - (k%len);
// //         //make a loop by attaching last to first
// //         cur.next = head;

// //         while(k > 0){
// //             cur=cur.next;
// //             k--;
// //         } 
// //         //get the head and break the link
// //         head = cur.next;
// //         cur.next = null;

// //         return head;

// //     }
// // }

File: 61-rotate-list/rotate-list.cpp
// class Solution {
// public:
//     int size(ListNode* curr){
//         int len=0;
//         while(curr!=NULL){
//             curr=curr->next;
//             len++;
//         }
//         return len;
//     }
//     ListNode* rotateRight(ListNode* head, int k) {
//         if(head==NULL){
//             return NULL;
//         }
//         int n = size(head);
//         int len=k%n;
//         len=n-len;
//         if(n==1||len==n){
//             return head;
//         }
//         int j=0;
//         ListNode *prev=head,*firstaddr=head;
//         while(prev!=NULL){
//             j++;
//             if(j==len){
//                 firstaddr=prev->next;
//                 prev->next=NULL;
//                 break;
//             }
//             prev=prev->next;
//         }
//         prev=firstaddr;
//         while(prev->next!=NULL){
//             prev=prev->next;
//         }
//         prev->next=head;
//         return firstaddr;
//     }
// };
class Solution {
    public:
        ListNode* rotateRight(ListNode* head, int k) {
        if(!head || !head->next || k==0) return head;
        //find length of LL
        ListNode*curr=head;
        int len=1;
        while(curr->next!=NULL && ++len){
            curr=curr->next;
        }
        ///go till that node
        curr->next=head; //make last node(curr) point to head-make circular
        k=k%len;
        k=len-k;
        while(k--){
            curr=curr->next;
        }
        head=curr->next;
        curr->next=NULL;
        return head;
    }
};
// // class Solution {
// // public:
// //     ListNode* rotateRight(ListNode* head, int k) {
// //        if(!head)
// //            return NULL;
// //         int n = 1;
// //         ListNode* tail = head;
// //         while(tail->next){
// //             n++;
// //             tail = tail->next;
// //         }
// //         tail->next = head;
// //         int size = n - k % n;
// //         for(int i = 0; i < size; i++){
// //             tail = tail->next;
// //         }  
// //         head = tail->next;
// //         tail->next = NULL;
// //         return head;
// //     }
// // };